chronos/asyncmacro2.nim

```diff
@@ -236,35 +285,87 @@ proc asyncSingleProc(prc: NimNode): NimNode {.compileTime.} =
     # -> return resultFuture
     outerProcBody.add newNimNode(nnkReturnStmt, prc.body[^1]).add(retFutureSym)
 
-  if prc.kind != nnkLambda: # TODO: Nim bug?
+  if prc.kind != nnkLambda and prc.kind != nnkProcTy: # TODO: Nim bug?
     prc.addPragma(newColonExpr(ident "stackTrace", ident "off"))
 
+  # The proc itself can't raise
+  prc.addPragma(nnkExprColonExpr.newTree(
+    newIdentNode("raises"),
+    emptyRaises))
+
+  prc.addPragma(newIdentNode("asyncinternal"))
+
   # See **Remark 435** in this file.
   # https://github.com/nim-lang/RFCs/issues/435
   prc.addPragma(newIdentNode("gcsafe"))
   result = prc
 
-  if subtypeIsVoid:
-    # Add discardable pragma.
-    if returnType.kind == nnkEmpty:
-      # Add Future[void]
-      result.params[0] =
-        newNimNode(nnkBracketExpr, prc)
-        .add(newIdentNode("Future"))
-        .add(newIdentNode("void"))
   if procBody.kind != nnkEmpty:
     result.body = outerProcBody
   #echo(treeRepr(result))
   #if prcName == "recvLineInto":
   #  echo(toStrLit(result))
 
+macro checkFutureExceptions(f, typ: typed): untyped =
+  # For RaiseTrackingFuture[void, (ValueError, OSError), will do:
+  # if isNil(f.error): discard
+  # elif f.error of type CancelledError: raise cast[ref CancelledError](f.error)
+  # elif f.error of type ValueError: raise cast[ref ValueError](f.error)
+  # elif f.error of type OSError: raise cast[ref OSError](f.error)
+  # else: raiseAssert("Unhandled future exception: " & f.error.msg)
+  #
+  # In future nim versions, this could simply be
+  # {.cast(raises: [ValueError, OSError]).}:
+  #   raise f.error
+  let e = getTypeInst(typ)[2]
+  let types = getType(e)
+
+  if types.eqIdent("void"):
+    return quote do:
+      if not(isNil(`f`.error)):
+        if `f`.error of type CancelledError:
+          raise cast[ref CancelledError](`f`.error)
+        else:
+          raiseAssert("Unhandled future exception: " & `f`.error.msg)
+
+  expectKind(types, nnkBracketExpr)
+  expectKind(types[0], nnkSym)
+  assert types[0].strVal == "tuple"
+  assert types.len > 1
+
+  result = nnkIfExpr.newTree(
+    nnkElifExpr.newTree(
+      quote do: isNil(`f`.error),
+      quote do: discard
+    )
+  )
+
+  result.add nnkElifExpr.newTree(
+    quote do: `f`.error of type CancelledError,
```

Comment:
why does this need special casing? ie if a function uses a call that potentially can be cancelled and doesn't handle that cancellation, it should `asyncraises: [CancelledError]`

Reply:
conversely, a call that disarms cancellederror will no longer leak cancellations and should have this effect

Reply:
I think it would add more noise than anything else (especially since we can't `push` this)
ie:
```nim
proc a {.async, asyncraises: [CancelledError].} =
  await sleepAsync(10.minutes)

asyncSpawn a()
```
In this setup, `a` can't actually raise cancelled error, but it's still in the tracking. You would have to do
```nim
proc a {.async, asyncraises: [].} =
  try:
    await sleepAsync(10.minutes)
  except CancelledError as exc:
   discard

asyncSpawn a()
```

And that would be the only case where it would make sense to track it: 95% of procedures will just bubble up the cancellation, after possibly cleaning some stuff
So I see little gain in tracking it, but a lot of noise

Reply:
The main problem here is that `a` can and indeed does raise a `CancelledError`, depending on what the caller does - it's just that it so happens that this functionality is not used by the caller in this particular case. 

There's also another thing to consider: some tasks are inherently not cancellable:

```nim
proc f() {.async.} = discard
```

can indeed never raise `CancelledError` because it will always returns a completed future (and cancelling a completed future has no effect) - putting cancellederror in asyncraises documents / makes explicit which tasks indeed are cancellable. 

We *should* be able to track this through `await` calls as well - ie intuitively `proc g() {.async, ..}= await f()` should be able to determine that no cancellation is possible.

We might need language to define non-cancellable tasks, but that's an orthogonal issue.

Eventually, we also want to fix the `push` issue, perhaps with some help from the compiler.

I *think* the right thing to do here is to make it explicit, at least for now - if we add the magic later on, it should be harmless - ie we must always support `asyncraises: [CancelledError]` - it should not be "wrong" to spell it out.

Reply:
> makes explicit which tasks indeed are cancellable

I don't fully agree here.
```nim
proc dial(address: MultiAddress): Future[Connection] {.async, asyncraises: [DialError, UpgradeError, CancelledError].} =
  let conn = await whatever(address)
  try:
    await conn.upgrade()
  except UpgradeError as exc:
    await conn.close()
    raise exc
  return conn
```
This procedure is technically cancellable, but it's wrong: conn will leak if we get cancelled during the upgrade
So this is not a good signal that the CancelledError is handled properly


> We should be able to track this through await calls as well - ie intuitively proc g() {.async, ..}= await f() should be able to determine that no cancellation is possible.

```nim
proc g() {.async, ...} = await asyncSleep(10.minutes)

g().cancel()
```
`await` can't determine if it's cancellable or not, since it depends on the behavior of callers & callees

> I think the right thing to do here is to make it explicit, at least for now

Well, that would be the only way to change it in the future without backward compatibility break, indeed. But still not convinced it makes sense in the first place

Reply:
> This procedure is technically cancellable, but it's wrong:

So think about it this way: 
* `dial` is cancellable because `upgrade` is cancellable (because ultimately, it passes though `select` which goes through the event loop which introduces the possibility for cancellation)
* how does the compiler know `upgrade` is cancellable? because it has a `CancelledError` raises effect
* when you're writing `dial`, the first thing you type is `asyncraises: []` assuming nothing is cancellable and there are no errors (with `push` support, this is what we'll enforce in every module top-level, similar to `push raises: []`)
* the compiler complains: `upgrade` raises cancellederror that you need to handle (together with a bunch of other errors)
* you now write the EH for cancellederror and reraise it

the shorter version of the above is that as you're typing `dial... asyncraises: [CancelledError]`, you have a clear and visual indication that cancellationerror is possible, and that you need to handle it in your function. One  can imagine a warning for functions that have overly broad raises lists as well (though this is dubious: a broad exception list is one way to make API future-proof, by including the possibility to raise new errors, even if the code presently doesn't).

In the hidden version, we're exactly where we are today: there's no indication that cancellation errors may happen, so programmers don't deal with it, just like they don't deal with exceptions in general. It also leaves no way to indicate that a function does _not_ support cancellation.


> await can't determine if it's cancellable or not, since it depends on the behavior of callers & callees

You're confusing whether a function is cancellable with whether it gets cancelled. The former is determined by the implementation of the function: `sleepAsync` is cancellable because it interacts with the event loop, and therefore should have a `raises: [CancelledError]`.

Consider this:
```
proc f(b: bool) {. raises [ValueError].} =
  if b:
    raise (ref ValueError)()
```

In the above case, the _caller_ determines whether the function actually raises (dynamically) - the same situation applies to cancellation and its associated exception: the caller might call cancel or not, but in the implementation of `f`, you have to assume that it's a possibility and write the cancellation handling code.

Reply:
> t, but in the implementation of f, you have to assume that it's a possibility and write the cancellation handling code.

the exception to this is if you know that you call no cancellable functions: then, you can safely write `f` without considering `cancel` - and have a `raises: []`.

By extension, `await` can also know if the entire chain of functions is cancellable or not by examining the raises effect of each called function - interestingly, this also tells you if a function actually could be made `sync` :)

Reply:
> sleepAsync is cancellable because it interacts with the event loop

Yes, but in practice, every async procedure will go to the event loop, either directly, or through it's callees. Because otherwise, why is it async? Which is my original point: almost every async procedure will raise the CancelledError, because almost every async procedure uses await on something eventually going to the event loop
The only (sensible) async procedures that don't raise CancelledError are the one swallowing it (which is dubious in other ways)


Your first scenario could work well if we forced the user to remap the CancelledError to something else, eg:
```nim
proc f {.raises [CancelledError].} =
  try: await sleepAsync(xx) # will throw ChildCancelledError
  except ChildCancelledError as exc: bubbleup(exc) # or whatever, a remap to CancelledError
```

Reply:
> Because otherwise, why is it async?

future-proofing, uniterruptible tasks, remapping CancelledError to a "regular" response

> one swallowing it (which is dubious in other ways)

well, one way to deal with cancellation is to give a non-cancellation error - ie change `CancelledError` to `0 bytes read`, breaking the cancellation chain: this gives functions that do go through the event loop a way to stay cancellation-free, isolating cancellation to specific parts of the code (just like you simplify other error codes generally: FileReadError becomes DbError becomes sometihng else as you move through layers)

Reply:
That could become interesting, and even a possible solution to the "cancellation race issue", but that will imply some heavy changes to the cancellation scheme

But ok, I'm starting to see some interesting possible behaviors, so I'll err on the "can be changed in a non-breaking way" side :)

Reply:
> Your first scenario could work well if we forced the user to remap the CancelledError to something else, eg:

I'd actually prefer this, except that it's insanely verbose. We have a less verbose solution for that though, `Result` :) Once the exception tracking is in, I'd say an interesting area to explore is Future/Result integration at some level (ranging from making `?` compatible to 

All of this discussion goes back to what information should be visible / captured, and where: the "true" return type of a cancellable `async` function, if you consider the full gamut of control-flow options, can be expressed roughly as `Future[Option[Either[T, XxxError]]`: through experience, we know that hiding `XxxError` from that signature leads to code not considering it - you're saying we should hide the `Option` part as well: the same thing happens.

Once you put both cancellation and error information in the function signature (via `asyncraises: [Cancelled, Xxx]`), the remaining practical differences between `Result` and Exception are:

* `Result` enforces call-site annotations (its main feature) - Exceptions bubble up without any visual indication at the call site
* `Result` allows static typing / type analysis (and permits dynamic typing) - `Exception` forces dynamic typing

We'll see how far the balance goes, but a good starting point nonetheless is to make the function signature rich(er): each library can then decide about call-site annnotations, as we build out the capabilities of these.

Perhaps we'll end up with `asynccancellable: [A, B]` to shorten things, or a `read(v: Future): Opt[Result[T, E]]` with nice syntactic sugar to make cancellation and errors easy to work with.

Reply:
d5eed32